reconcile.go

```diff
@@ -0,0 +1,264 @@
+	// stable before doing the transfer
+	return demoted, nil
+}
+
+// getFailedServers aggregates all of the information about servers that the consuming application believes are in
+// a failed/left state (indicated by the NodeStatus field on the Server type) as well as stale servers that are
+// in the raft configuration but not know to the consuming application. This function will do nothing with
+// that information and is purely to collect the data.
+func (a *Autopilot) getFailedServers() (*FailedServers, int, error) {
+	conf := a.delegate.AutopilotConfig()
+	if conf == nil || !conf.CleanupDeadServers {
+		return nil, 0, nil
+	}
+
+	staleRaftServers := make(map[raft.ServerID]raft.Server)
+	raftConfig, err := a.getRaftConfiguration()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Populate a map of all the raft servers. We will
+	// remove some later on from the map leaving us with
+	// just the stale servers.
+	var voters int
+	for _, server := range raftConfig.Servers {
+		staleRaftServers[server.ID] = server
+
+		if server.Suffrage == raft.Voter {
+			voters++
+		}
+	}
+
+	var failed FailedServers
+	for id, srv := range a.delegate.KnownServers() {
+		raftSrv, found := staleRaftServers[id]
+		if found {
+			delete(staleRaftServers, id)
+		}
+
+		if srv.NodeStatus == NodeFailed {
+			if found && raftSrv.Suffrage == raft.Voter {
+				failed.FailedVoters = append(failed.FailedVoters, srv)
+			} else if found {
+				failed.FailedNonVoters = append(failed.FailedNonVoters, srv)
+			}
+		}
+	}
+
+	for id, srv := range staleRaftServers {
+		if srv.Suffrage == raft.Voter {
+			failed.StaleVoters = append(failed.StaleVoters, id)
+		} else {
+			failed.StaleNonVoters = append(failed.StaleNonVoters, id)
+		}
+	}
+
+	return &failed, voters, nil
+}
+
+// pruneDeadServers will find stale raft servers and failed servers as indicated by the consuming application
+// and remove them. For stale raft servers this means removing them from the Raft configuration. For failed
+// servers this means issuing RemoveFailedNode calls to the delegate. All stale/failed non-voters will be
+// removed first. Then stale voters and finally failed servers. For servers with voting rights we will
+// cap the number removed so that we do not remove too many at a time and do not remove nodes to the
+// point where the number of voters would be below the MinQuorum value from the autopilot config.
+// Additionally the delegate will be consulted to determine if all of the removals should be done and
+// can filter the failed servers listings if need be.
+func (a *Autopilot) pruneDeadServers() error {
+	conf := a.delegate.AutopilotConfig()
+	if conf == nil || !conf.CleanupDeadServers {
+		return nil
+	}
+
+	failed, voters, err := a.getFailedServers()
+	if err != nil || failed == nil {
+		return err
+	}
+
+	failed = a.promoter.FilterFailedServerRemovals(a.delegate.AutopilotConfig(), a.GetState(), failed)
```

Comment:
Can we just use `conf` here instead of re-calling `AutopilotConfig()`?

This makes think about your points around the issue of inconsistent `*Config`s and whether we should use a push-based model of updating the `*Config` instead so that we can't get ourselves into that situation at all.

Another alternative might be to only call `AutopilotConfig()` in the `run()` loop and pass handles down to `reconcile` and `pruneDeadServers` to ensure that the same config is used for the entirety of each loop.

Reply:
oh, yes. this is definitely a mistake.